listenbrainz/webserver/views/user_timeline_event_api.py

```diff
@@ -85,3 +96,196 @@ def create_user_recording_recommendation_event(user_name):
     event_data['created'] = event_data['created'].timestamp()
     event_data['event_type'] = event_data['event_type'].value
     return jsonify(event_data)
+
+
+@user_timeline_event_api_bp.route('/user/<user_name>/feed/events', methods=['OPTIONS', 'GET'])
+@crossdomain(headers="Authorization, Content-Type")
+@ratelimit()
+def user_feed(user_name: str):
+    """ Get feed events for a user's timeline.
+
+    :param user_name: The MusicBrainz ID of the user whose timeline is being requested.
+    :type user_name: ``str``
+    :param max_ts: If you specify a ``max_ts`` timestamp, events with timestamps less than or equal to it will be returned
+    :param min_ts: If you specify a ``min_ts`` timestamp, events with timestamps greated than or equal to it will be returned
+    :param count: Optional, number of events to return. Default: :data:`~webserver.views.api.DEFAULT_ITEMS_PER_GET` . Max: :data:`~webserver.views.api.MAX_ITEMS_PER_GET`
+    :statuscode 200: Successful query, you have feed events!
+    :statuscode 400: Bad request, check ``response['error']`` for more details.
+    :statuscode 401: Unauthorized, you do not have permission to view this user's feed.
+    :statuscode 404: User not found
+    :resheader Content-Type: *application/json*
+    """
+
+    user = validate_auth_header()
+    if user_name != user['musicbrainz_id']:
+        raise APIUnauthorized("You don't have permissions to view this user's timeline.")
+
+    db_conn = webserver.create_timescale(current_app)
+    min_ts, max_ts, count, time_range = _validate_get_endpoint_params(db_conn, user_name)
+    if min_ts is None and max_ts is None:
+        max_ts = int(time.time())
+
+    users_following = db_user_relationship.get_following_for_user(user['id'])
+    if len(users_following) == 0:
+        return jsonify({'payload': {
+            'count': 0,
+            'user_id': user_name,
+            'events': [],
+        }})
+
+    # get all listen events
+    musicbrainz_ids = [user['musicbrainz_id'] for user in users_following]
+    listen_events = get_listen_events(db_conn, musicbrainz_ids, min_ts, max_ts, count, time_range)
+
+    # for events like "follow" and "recording recommendations", we want to show the user
+    # their own events as well
+    users_for_feed_events = users_following + [user]
+    follow_events = get_follow_events(
+        user_ids=tuple(user['id'] for user in users_for_feed_events),
+        min_ts=min_ts or 0,
+        max_ts=max_ts or int(time.time()),
+        count=count,
+    )
+
+    recording_recommendation_events = get_recording_recommendation_events(
+        users_for_events=users_for_feed_events,
```

Comment:
These blocks that return empty data bug me. Duplicate code; can we just return None or at least a helper function?

Reply:
Not completely sure I understand, this code doesn't return empty data, it returns recording recommendation events.